Cubical/Algebra/RingSolver/CommRingSolver.agda

```diff
@@ -0,0 +1,221 @@
+        ⟦ e ⟧ (x ∷ xs)
+        + eval (ℕ.suc n) (normalize (ℕ.suc n) e₁) (x ∷ xs)
+      ≡⟨ cong (λ u → ⟦ e ⟧ (x ∷ xs) + u)
+              (isEqualToNormalform (ℕ.suc n) e₁ (x ∷ xs)) ⟩
+        ⟦ e ⟧ (x ∷ xs) + ⟦ e₁ ⟧ (x ∷ xs) ∎
+
+  isEqualToNormalform ℕ.zero (e ·' e₁) [] =
+        eval ℕ.zero (normalize ℕ.zero e ·ₕ normalize ℕ.zero e₁) []
+      ≡⟨ ·Homeval ℕ.zero (normalize ℕ.zero e) _ [] ⟩
+        eval ℕ.zero (normalize ℕ.zero e) []
+        · eval ℕ.zero (normalize ℕ.zero e₁) []
+      ≡⟨ cong (λ u → u · eval ℕ.zero (normalize ℕ.zero e₁) [])
+              (isEqualToNormalform ℕ.zero e []) ⟩
+        ⟦ e ⟧ []
+        · eval ℕ.zero (normalize ℕ.zero e₁) []
+      ≡⟨ cong (λ u → ⟦ e ⟧ [] · u) (isEqualToNormalform ℕ.zero e₁ []) ⟩
+        ⟦ e ⟧ [] · ⟦ e₁ ⟧ [] ∎
+
+  isEqualToNormalform (ℕ.suc n) (e ·' e₁) (x ∷ xs) =
+        eval (ℕ.suc n) (normalize (ℕ.suc n) e
+                         ·ₕ normalize (ℕ.suc n) e₁) (x ∷ xs)
+      ≡⟨ ·Homeval (ℕ.suc n) (normalize (ℕ.suc n) e) _ (x ∷ xs) ⟩
+        eval (ℕ.suc n) (normalize (ℕ.suc n) e) (x ∷ xs)
+        · eval (ℕ.suc n) (normalize (ℕ.suc n) e₁) (x ∷ xs)
+      ≡⟨ cong (λ u → u · eval (ℕ.suc n) (normalize (ℕ.suc n) e₁) (x ∷ xs))
+              (isEqualToNormalform (ℕ.suc n) e (x ∷ xs)) ⟩
+        ⟦ e ⟧ (x ∷ xs)
+        · eval (ℕ.suc n) (normalize (ℕ.suc n) e₁) (x ∷ xs)
+      ≡⟨ cong (λ u → ⟦ e ⟧ (x ∷ xs) · u)
+              (isEqualToNormalform (ℕ.suc n) e₁ (x ∷ xs)) ⟩
+        ⟦ e ⟧ (x ∷ xs) · ⟦ e₁ ⟧ (x ∷ xs) ∎
+
+  solve :
+    {n : ℕ} (e₁ e₂ : ℤExpr n) (xs : Vec (fst R) n)
+    (p : eval n (normalize n e₁) xs ≡ eval n (normalize n e₂) xs)
+    → ⟦ e₁ ⟧ xs ≡ ⟦ e₂ ⟧ xs
+  solve e₁ e₂ xs p =
+    ⟦ e₁ ⟧ xs                  ≡⟨ sym (isEqualToNormalform _ e₁ xs) ⟩
+    eval _ (normalize _ e₁) xs ≡⟨ p ⟩
+    eval _ (normalize _ e₂) xs ≡⟨ isEqualToNormalform _ e₂ xs ⟩
+    ⟦ e₂ ⟧ xs ∎
+
+ℤExpr : (R : CommRing {ℓ}) (n : ℕ)
+        → _
+ℤExpr R n = EqualityToNormalform.ℤExpr R n
+
+solve : (R : CommRing {ℓ})
+        {n : ℕ} (e₁ e₂ : ℤExpr R n) (xs : Vec (fst R) n)
+        (p : eval n (EqualityToNormalform.normalize R n e₁) xs ≡ eval n (EqualityToNormalform.normalize R n e₂) xs)
+        → _
+solve R = EqualityToNormalform.solve R
+
+module VarNames3 (R : CommRing {ℓ}) where
+  X1 : ℤExpr R 3
+  X1 = ∣ Fin.zero
+
+  X2 : ℤExpr R 3
+  X2 = ∣ (suc Fin.zero)
+
+  X3 : ℤExpr R 3
+  X3 = ∣ (suc (suc Fin.zero))
+
+module VarNames4 (R : CommRing {ℓ}) where
+  X1 : ℤExpr R 4
+  X1 = ∣ Fin.zero
+
+  X2 : ℤExpr R 4
+  X2 = ∣ (suc Fin.zero)
+
+  X3 : ℤExpr R 4
+  X3 = ∣ (suc (suc Fin.zero))
+
+  X4 : ℤExpr R 4
+  X4 = ∣ (suc (suc (suc Fin.zero)))
+
+module VarNames5 (R : CommRing {ℓ}) where
+  X1 : ℤExpr R 5
+  X1 = ∣ Fin.zero
+
+  X2 : ℤExpr R 5
+  X2 = ∣ (suc Fin.zero)
+
+  X3 : ℤExpr R 5
+  X3 = ∣ (suc (suc Fin.zero))
+
+  X4 : ℤExpr R 5
+  X4 = ∣ (suc (suc (suc Fin.zero)))
+
+  X5 : ℤExpr R 5
+  X5 = ∣ (suc (suc (suc (suc Fin.zero))))
+
+module VarNames6 (R : CommRing {ℓ}) where
```

Comment:
Couldn't we just have this parametrized by a natural number `n` and have `X_ : N -> ZExpr R n`? (defined by `X_ x = | x % n` where `x % n` is x modulo n viewed as an element of Fin n

Reply:
PS: I don't think the `%` operation is defined anywhere, but it should be too hard to add (I hope)

Reply:
Maybe I should merge this as it is and we make an issue for @felixwellen ? Would be nice if @mzeuner could experiment a bit with simplifying more proofs about localization

Reply:
Yes. 
I tried what you suggested before making the definitions, but annoyingly and contrary to my intuition, the 'n' in 'Fin n' could not be left implicit. But I could try to figure out why it didn't work or obsolete the issue by building a reflection interface.

Reply:
Ah, wait. I didn't use modulo arithmetic when I tried. But in any case, I'm for putting it in a separate issue.

Reply:
Yeah, I don't know if the modulo approach is a good idea. It's probably just the old Coq programmer inside me trying to avoid using dependent types... Anyway I'll merge this now and as you say it's probably not worth having a nicer solution as this will be removed once we have some reflection.